### PR TITLE
Provide template files docs

### LINKED ISSUE #44: Generating date with proper format (or format generic dates)
Working on https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/43 I needed to create 2 dates with specific format: a ISO8601 compliant timestamp and a Syslog compliant timestamp.

In both cases (from my understanding) this cannot be expressed in our templates. 

In Elasticsearch `date` type fields can set a [`format`](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html) parameter to specify date format, but this is not supported in this tool (not sure it **should** be, as this parameter applies to index field mapping).

A viable solution would be to format dates in the template itself (separating generation and presentation, something @aspacca and me discussed and which we consider a bonus point of the template approach recently introduced).

This tool includes [Masterminds/sprig](https://github.com/Masterminds/sprig) templating functions, but only "pure" ones ([here](https://github.com/elastic/elastic-integration-corpus-generator-tool/blob/a86918989bd2dc8389a91bb597ec5b860961fa67/pkg/genlib/generator_with_text_template.go#L34) we are mapping only "hermetic functions"). [These functions](https://github.com/Masterminds/sprig/blob/581758eb7d96ae4d113649668fa96acc74d46e7f/functions.go#L70) **do not** include `date`.

Following this, I'm not sure how to move forward in the linked PR, or what is the suggested approach for custom dates in templates. Can we rethink the exclusion of non hermetic functions and add them too?

### PR DESCRIPTION
**Pull request by endorama**
This PR adds some documentation for template files.

Consider it a draft to be improved upon.

Closes #44

### CODE REVIEW

File: docs/writing-templates.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+- (_mandatory_) `fields.yml`: a field definition file
+- (_optional_) `configs.yml`: a field generation configuration file
+- (_mandatory_) `gotext.tpl`: a `gotext` template file
+- (_optional_) `placeholder.tpl`: a `placeholder` template file
+
+## `fields.uyml` - Field definitions
+
+A `YAML` file containing field mapping definitions. Ideally this file is extracted from Integration packages, but there is no automation for doing so at the moment.
+
+_NOTE_: how do we manage schema differences between various generation schemas (A, B, C, D)?
```
Comment (aspacca) [inline docs/writing-templates.md:37]: if we reference "schemas A, B, C, D", we must as well define them :)

**Review by aspacca — APPROVED** [review]

**Comment by endorama** [comment]
I added some more pages. @aspacca do you have at hand some benchmark data to put in the performances page? Otherwise I'll generate them.

**Comment by aspacca** [comment]
> do you have at hand some benchmark data to put in the performances page?

this is everything I have https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/40

File: docs/field-configurations.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+They must be added to a file named `configs.yml` in the assets template folder of a data stream.
+
+## Config entries definition
+
+The config file is a yaml file consisting of an array of config entry.
+
+For each config entry the following fields are available:
+- `name` *mandatory*: dotted path field, as in `fields.yml`
+- `fuzziness` *optional (`long` and `double` type only)*: delta from the previous generated value for the same field
```
Comment (aspacca) [inline docs/field-configurations.md:13]: let's remember to replicate here the information from https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/46/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R257-R259

Comment (endorama) [reply docs/field-configurations.md:13]: I'll wait for #46 to be closed before moving forward.

File: docs/writing-templates.md (hunk 1 above)
Comment (endorama) [reply docs/writing-templates.md:37]: definitions added to https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/45/commits/b834ef025288289c95e0ba30d748d1b39a48c29a

File: docs/writing-templates.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+- the use case is generating fake but plausible data, a challenge for a random data generator, as it implies:
+  - in observability dataset usually exhibit high value cardinality; generating high cardinality dataset is a goal, for more information see [cardinality.md](./cardinality.md);
+  - on the opposite side, a data generator may produce data with high similarity, which will not exhibit a behaviour similar to real data; while we aim to increase data dissimilarity, is not a goal to have complete real like data generation;
+  - another characteristic of datasets in observability is high field dimensionality; generating high dimensionality dataset is a goal, for more information see [dimensionality.md](./dimensionality.md);
+- data generation may not be a real time operation executed together/just before using the generated data; we aim to provide corpus of generated data that can be replayed as if they were being emitted in real time with the goal of allowing creation of big and complex datasets.
+
+## Folder structure
+
+All templates are contained in `/assets/templates` folder. Within it there should be a folder for each Integration data stream templates are available from.
```
Comment (aspacca) [inline docs/writing-templates.md:25]: oki with this naming and structure standard: let's just clarify that the name of the folder for the integration data stream must be in the format `fmt.Sprintf("%s.%s", packageName, dataStreamName)`

Comment (aspacca) [reply docs/writing-templates.md:25]: nit: you have both "schema A" and "schema B" assets on https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/43

I will propose to use a specific folder for each schema (like you have `schema-b` but not `schema-a`)

let's update the content of this section once we agree on a format

File: docs/writing-templates.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+- (_mandatory_) `fields.yml`: a field definition file
+- (_optional_) `configs.yml`: a field generation configuration file
+- (_mandatory_) `gotext.tpl`: a `gotext` template file
+- (_optional_) `placeholder.tpl`: a `placeholder` template file
+
+## `fields.uyml` - Field definitions
+
+A `YAML` file containing field mapping definitions. Ideally this file is extracted from Integration packages, but there is no automation for doing so at the moment.
+
+_NOTE_: how do we manage schema differences between various generation schemas (A, B, C, D)?
```
Comment (aspacca) [inline docs/writing-templates.md:37]: > how do we manage schema differences between various generation schemas (A, B, C, D)?

my proposal: https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/45#discussion_r1090338757

File: docs/dimensionality.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Dimensionality
+
+The dimensionality of a dataset is the number of fields that it has. **High dimensionality** datasets have many attributes.
+
+Dimensionality plays a role in `array` and `object` type fields.
+
+Dimensionality is valuable to test storage consumption, in particular for metrics, where the data point itself is small compared to the metadata enriching it.
+
+At the moment the number of attribute for these fields is random and is not configurable through Field generation configurations.
```
[the remark this answers is not in the document]
Comment (endorama) [reply docs/dimensionality.md:9]: That's great!!

**Comment by endorama** [comment]
Docs updated, asking for a second review

File: docs/writing-templates.md [hunk 6]
```diff
@@ -24,18 +24,20 @@ This operation has some caveats:
 
 All templates are contained in `/assets/templates` folder. Within it there should be a folder for each Integration data stream templates are available from.
 
-Within these folder templates files should be added:
+The folder MUST be named as: `<package name>.<data stream>`. For example: `apache.access`, `apache.error`, `aws.billin`, `aws.vpcflow`.
```
Comment (aspacca) [inline docs/writing-templates.md:27]: typo: `aws.billing`

File: docs/writing-templates.md [hunk 7]
```diff
@@ -24,18 +24,20 @@ This operation has some caveats:
...
+
+As this tool aims to support different [schemas](./data-schemas.md) for generated data, files related to a schema should be put within a folder named after the schema they relate to: `schema-a`, `schema-b`, `schema-c`, `schema-d`.
```
Comment (aspacca) [inline docs/writing-templates.md:29]: let's make an example:
```
For example the final folder where to place the files listed below will be `/assets/templates/apache.error/schema-a`
```

**Review by aspacca — APPROVED** [review]

**Review by aspacca — APPROVED** [review]

**Pull request merged**